07_StagingAndMultipleQueues/main.cpp

```diff
@@ -2,78 +2,580 @@
+		std::array<core::smart_refctd_ptr<nbl::video::IGPUCommandPool>, FRAMES_IN_FLIGHT> commandPools;
+		std::array<core::smart_refctd_ptr<nbl::video::IGPUCommandBuffer>, FRAMES_IN_FLIGHT> commandBuffers;
+		std::fill(commandPools.begin(), commandPools.end(), nullptr);
+		for (uint32_t i = 0u; i < FRAMES_IN_FLIGHT; ++i)
+		{
+			commandPools[i] = m_device->createCommandPool(transferUpQueue->getFamilyIndex(), commandPoolFlags);
+			commandPools[i]->createCommandBuffers(IGPUCommandPool::BUFFER_LEVEL::PRIMARY, std::span(commandBuffers.data() + i, 1), core::smart_refctd_ptr(m_logger));
+		}
 
-		// This time we will load images and compute their histograms and output them as CSV
-		bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+		// TODO: loop over images to load
+			// DONE: If imageIndex>=FRAMES_IN_FLIGHT need to block for Semaphore at value imageIndex+1-FRAMES_IN_FLIGHT
+			// DONE: Create IGPUImage with matching parameters, EXCEPT format, format we promote using IPHysicalDevice
+			// DONE: just use Transfer UP queue with IUtilities to fill image
+			// DONE: layout transition from TRANSFER_DST to SHADER_READ_ONLY/SAMPLED
+			// TODO: make sure to use the non-blocking variant of `IUtilities` to control the `IQueue::SSubmitInfo::signalSemaphores`
+
+		for (uint32_t imageIdx = 0; imageIdx < IMAGE_CNT; ++imageIdx)
 		{
-			// Remember to call the base class initialization!
-			if (!device_base_t::onAppInitialized(std::move(system)))
-				return false;
-			if (!asset_base_t::onAppInitialized(std::move(system)))
-				return false;
+			const size_t resourceIdx = imageIdx % FRAMES_IN_FLIGHT;
+			const auto& imageToLoad = imagesToLoad[imageIdx];
+			auto& cmdBuff = commandBuffers[resourceIdx]; 
 
-			// TODO: for later
-			// - fire up 2 aux threads (one to upload images and transfer ownership [if necessary], another to acquire ownership of histogram buffers and write out CSVs)
-			// - main thread grabs `IGPUImage` from a queue and acquires ownership [if necessary] performs all setup to launch a dispatch hands off a histogram buffer
+			auto isResourceReused = waitForResourceAvailability(m_imagesLoadedSemaphore.get(), imageIdx);
+			if(isResourceReused)
+				cmdBuff->reset(IGPUCommandBuffer::RESET_FLAGS::NONE);
 
-			return true;
+			IGPUImage::SCreationParams imgParams;
+			imgParams.type = IImage::E_TYPE::ET_2D;
+			imgParams.extent.height = cpuImages[0]->getCreationParameters().extent.height;
+			imgParams.extent.width = cpuImages[0]->getCreationParameters().extent.height;
+			imgParams.extent.depth = cpuImages[0]->getCreationParameters().extent.depth;
+			IPhysicalDevice::SImageFormatPromotionRequest formatPromotionRequest;
+			IPhysicalDevice::SFormatImageUsages::SUsage usage;
+			usage.sampledImage = 1;
+			usage.transferDst = 1;
+			formatPromotionRequest.usages = usage;
+			// why my images are VK_FORMAT_R8_SRGB?
```

Comment:
what do you mean?

Reply:
ok I see, the format promotion fails?

Reply:
seems like it

Reply:
ok I'll use this as a test and fix

Reply:
ok, it was my fault, i haven't set an `originalFormat` field of the `SImageFormatPromotionRequest` struct